### ISSUE TITLE
Dune sometimes changes *.opam files in release mode

### ISSUE THREAD

**Issue by glondu** [ocaml/dune#8417]
## Expected Behavior

In "reproducible  release  mode" (`-p` or `--release`), Dune should not change *.opam files.

## Actual Behavior

Dune sometimes changes these files.

## Reproduction

```
$ wget https://github.com/ocaml/graphics/releases/download/5.1.2/graphics-5.1.2.tbz
$ tar xf graphics-5.1.2.tbz
$ cd graphics-5.1.2
$ git init && git add . && git commit -m "Initial commit"
$ dune build -p graphics
$ git diff
diff --git a/graphics.opam b/graphics.opam
index 0f53f1e..1386301 100644
--- a/graphics.opam
+++ b/graphics.opam
@@ -1,6 +1,6 @@
-version: "5.1.2"
 # This file is generated by dune, edit dune-project instead
 opam-version: "2.0"
+version: "5.1.2"
 synopsis: "The OCaml graphics library"
 description: """
 The graphics library provides a set of portable drawing
@@ -38,4 +38,4 @@ build: [
     "@doc" {with-doc}
   ]
 ]
-dev-repo: "git+https://github.com/ocaml/graphics.git"
\ No newline at end of file
+dev-repo: "git+https://github.com/ocaml/graphics.git"
```

Note that I took graphics as an example, but it happens in many other packages.

## Specifications

- Version of `dune` (output of `dune --version`): 3.9.1 (`dune --version` actually returns `n/a`)
- Version of `ocaml` (output of `ocamlc --version`): 4.13.1
- Operating system (distribution and version): Debian sid (as of 2023-08-18)

## Context

This was discovered by Debian QA, which checks that sources are not modified during build. [Debian bug](https://bugs.debian.org/1050003).

**Comment by glondu** [comment]
This problem is not fixed with 3.11.x, actually. The original repro case still applies.

**Comment by Alizter** [comment]
I believe this was closed with #14108. From dune lang 3.23 onwards, opam files will be diffed rather than autopromoted. Furthermore in release mode they will not be diffed at all.

**Comment by glondu** [comment]
I just tried dune 3.23 with easy-format 1.3.4, and `dune build -p easy-format` still alters easy-format.opam...

**Comment by Alizter** [comment]
@glondu Does `easy-format` use `dune lang 3.23`. The dune version will not be enough, since we version behaviour changes like this. You could try patching dune lang in the dune-project file to 3.23 on the debian packaging side and try building it that way.

**Comment by glondu** [comment]
> [@glondu](https://github.com/glondu) Does `easy-format` use `dune lang 3.23`.

No. (BTW, dune 3.23 is not yet packaged in Debian, because of a conflict yet to be investigated.)

> The dune version will not be enough, since we version behaviour changes like this.

`easy-format` is not the only package affected by this buggy behaviour (see https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=1050003 and all related bug listed in "Blocking fix for").

> You could try patching dune lang in the dune-project file to 3.23 on the debian packaging side and try building it that way.

Changing `dune lang` in dune-project is upstream's responsibility, not Debian's. Besides, bumping it on the "debian packaging side" could lead to other behavioural changes, which could lead to further patching... This is definitely something we don't want.

I would have preferred to see the buggy behaviour fixed without `dune lang` version bump, but I understand if you disagree, in which case, you can close this issue.

**Comment by Alizter** [comment]
@glondu Thanks for taking the time to explain your perspective. I've reopened the issue for now until we can find a decent solution. Off the top of my head, we have tried changing this behaviour previously, but users have complained that they relied(?!) on it. I'll get back to this issue when I have the time.
